File: 12015.js
const fs = require('fs');

function solve() {
	let input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;

	const N = +input[inputIdx++];
	const arr = input[inputIdx++].split(' ').map((v) => +v);

	let dp = [arr[0]];

	const bisectLeft = (x) => {
		let mid,
			lo = 0,
			hi = dp.length;
		while (lo < hi) {
			mid = Math.floor((lo + hi) / 2);
			dp[mid] < x ? (lo = mid + 1) : (hi = mid);
		}
		return lo;
	};

	for (let i = 1; i < N; i++) {
		if (dp[dp.length - 1] < arr[i]) dp.push(arr[i]);
		else dp[bisectLeft(arr[i])] = arr[i];
	}
	return dp.length;
}

console.log(solve());

File: 12015.py
from sys import stdin
from bisect import bisect_left

input = stdin.readline


def solve():
    N = int(input())
    arr = [*map(int, input().split())]
    dp = [arr[0]]

    for i in range(1, N):
        if arr[i] > dp[-1]:
            dp.append(arr[i])
        else:
            dp[bisect_left(dp, arr[i])] = arr[i]

    return len(dp)


if __name__ == '__main__':
    print(solve())
